python/ray/data/_internal/arrow_ops/transform_pyarrow.py

```diff
@@ -57,33 +58,180 @@ def sort(table: "pyarrow.Table", sort_key: "SortKey") -> "pyarrow.Table":
+
+    Args:
+        table: The input pyarrow Table
+        indices: Row indices to select, can be:
+            - List[int]: Will be converted to numpy array
+            - numpy.ndarray
+            - pyarrow Array
+            - pyarrow ChunkedArray
+
+    Returns:
+        A new pyarrow Table with the selected rows
     """
-    from ray.air.util.transform_pyarrow import (
-        _concatenate_extension_column,
-        _is_column_extension_type,
-    )
+    from ray.air.util.transform_pyarrow import _is_column_extension_type
 
-    if any(_is_column_extension_type(col) for col in table.columns):
-        new_cols = []
-        for col in table.columns:
-            if _is_column_extension_type(col) and col.num_chunks > 1:
-                # .take() will concatenate internally, which currently breaks for
-                # extension arrays.
-                col = _concatenate_extension_column(col)
-            new_cols.append(col.take(indices))
-        table = pyarrow.Table.from_arrays(new_cols, schema=table.schema)
-    else:
-        table = table.take(indices)
-    return table
+    # Convert indices to numpy array if list
+    if isinstance(indices, list):
+        indices = np.array(indices, dtype=np.int64)
+
+    # After this point, indices is guaranteed to be ndarray/Array/ChunkedArray
+    indices_len = len(indices)
+    new_cols = []
+
+    for col in table.columns:
+        needs_chunking = False
+        if _is_column_extension_type(col):
+            # Handle extension arrays (like ArrowTensorArray) specially:
+            # 1. If it has multiple chunks, combine them since extension arrays
+            #    often break with internal concatenation
+            # 2. Only check for index overflow since extension arrays don't use
+            #    Arrow's internal offset arrays
+            if isinstance(col, pyarrow.ChunkedArray) and col.num_chunks > 1:
+                col = combine_chunked_array(col)
+            needs_chunking = _check_max_offset(indices, indices_len)
+        elif isinstance(col, pyarrow.ChunkedArray):
+            # For regular ChunkedArrays, we need chunking in two cases:
+            # 1. If indices would cause offset overflow
+            # 2. If concatenating the chunks would exceed 2GB (Arrow's 32-bit limit)
+            needs_chunking = _check_max_offset(indices, indices_len) or (
+                col.num_chunks > 1
+                and sum(chunk.nbytes for chunk in col.chunks) > MAX_INT32
+            )
+        else:
+            # For regular Arrays, we only need to check index overflow
+            needs_chunking = _check_max_offset(indices, indices_len)
+
+        if needs_chunking:
+            # Process indices in batches to stay within Arrow's limits
+            new_col = _process_large_indices(col, indices, indices_len)
+        else:
+            # Take all indices at once since we're within safe limits
+            new_col = pc.take(col, indices)
+
+        new_cols.append(new_col)
```

Comment:
All of that is already handled in `_try_combine_chunks_safe`, let's use that instead